### PR TITLE
Namespace context

### PR DESCRIPTION
**Pull request by hojberg**
## Overview
Add a `Perspective` type to hold the current codebase branch allowing
for nicer definition routes, namespace specific routes, and definitions
within namespace specific routes such as
`@codebase-hash/terms/base/List/map`,
`/@codebase-hash/namespaces/base/List`, and
`/@codebase-hash/namespaces/base/List/-/terms/map` respectively.

This adds support for those routes in the `Route` module, but all links
still go to absolute definition urls such as
`latest/types/@base.List.map-hash`.

(url parts prefixed with `@` in these examples represent Unison hashes)

Part of https://github.com/unisonweb/codebase-ui/issues/78

## Loose ends
This depends on https://github.com/unisonweb/unison/pull/2283

### CODE REVIEW

File: src/Api.elm [hunk 1]
```diff
@@ -10,37 +10,36 @@ module Api exposing
...
 
 
 type Endpoint
     = Endpoint (List String) (List QueryParameter)
 
 
-type ApiRequest a msg
-    = ApiRequest Endpoint (Decode.Decoder a) (Result Http.Error a -> msg)
+list : PerspectiveParams -> String -> Endpoint
+list perspectiveParams fqnOrHash =
```
Comment (hojberg) [inline src/Api.elm:37]: Various endpoints now take a `Perspective` or (as in the case of list) a `PerspectiveParams` to ensure data is fetched within the right context.

Comment (hojberg) [reply src/Api.elm:37]: (The list API does not actually support rootBranch and relativeTo right now, but is intended to. Sending the params doesn't do anything as they are just ignored).

File: src/App.elm [hunk 2]
```diff
@@ -51,15 +50,9 @@ type alias Model =
     }
 
 
-init : Flags -> Url -> Nav.Key -> ( Model, Cmd Msg )
-init flags url navKey =
+init : Env -> Route -> Nav.Key -> ( Model, Cmd Msg )
+init env route navKey =
```
Comment (hojberg) [inline src/App.elm:54]: App is being wrapped by the new Main module that ensures there's a codebase hash and a perspective before starting up the app.

File: src/Main.elm [hunk 3]
```diff
@@ -0,0 +1,136 @@
...
+    case model of
+        Initializing _ ->
+            { title = "Loading.."
+            , body = [ div [] [ text "Loading.." ] ]
+            }
+
+        InitializationError _ _ ->
+            { title = "Application Error"
+            , body = [ div [] [ text "Application Error" ] ]
+            }
```
Comment (hojberg) [inline src/Main.elm:127]: This is visible at the very edge if the application fails to load completely. We should improve the styling and messaging here. Tracked here #186

**Review by pchiusano — APPROVED** [review]
👍

**Pull request merged**